### PR TITLE
Arithmetic functions implementation for EVM->EWasm translator.

### PR DESCRIPTION
**Pull request by rodiazet**
### Description

Add div, sdiv, mod, smod, exp, lt, sar, addmod, mulmod, signextend

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### CODE REVIEW

**Comment by axic** [comment]
Also blocked on the same problem as #8049: string literal too big for msvc.

**Comment by rodiazet** [comment]
> Also blocked on the same problem as #8049: string literal too big for msvc.

I pushed fix for this issue. It works but let me know if it's acceptable for you.

**Comment by axic** [comment]
What is the fix?

**Comment by rodiazet** [comment]
> What is the fix?

Starting at line 615 EVMToEwasmTranslator.cpp

**Comment by rodiazet** [comment]
@axic @chriseth Should I do something more with this PR?

File: libyul/backends/wasm/EVMToEwasmTranslator.cpp [hunk 1]
```diff
@@ -124,36 +151,338 @@ function mul_128x128_256(x1, x2, y1, y2) -> r1, r2, r3, r4 {
...
 function div(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
-	// TODO implement properly
-	r4 := i64.div_u(x4, y4)
+	// Based on https://github.com/ewasm/evm2wasm/blob/master/wasm/DIV.wast
+	if iszero256(y1, y2, y3, y4) {
+		r1 := 0
+		r2 := 0
+		r3 := 0
+		r4 := 0
+	}
```
Comment (axic) [inline libyul/backends/wasm/EVMToEwasmTranslator.cpp:250]: Why not use `leave` here and then there's no need for the second if?

File: libyul/backends/wasm/EVMToEwasmTranslator.cpp [hunk 2]
```diff
@@ -124,36 +151,338 @@ function mul_128x128_256(x1, x2, y1, y2) -> r1, r2, r3, r4 {
...
 function mod(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
-	// TODO implement properly
-	r4 := i64.rem_u(x4, y4)
+ 	// Based on https://github.com/ewasm/evm2wasm/blob/master/wasm/MOD.wast
+	if iszero256(y1, y2, y3, y4) {
+		r1 := 0
+		r2 := 0
+		r3 := 0
+		r4 := 0
+	}
```
Comment (axic) [inline libyul/backends/wasm/EVMToEwasmTranslator.cpp:311]: Same here.

File: libyul/backends/wasm/EVMToEwasmTranslator.cpp [hunk 3]
```diff
@@ -124,36 +151,338 @@ function mul_128x128_256(x1, x2, y1, y2) -> r1, r2, r3, r4 {
...
+}
+function mod320(x1, x2, x3, x4, x5, y1, y2, y3, y4, y5) -> r1, r2, r3, r4, r5 {
+	// Based on https://github.com/ewasm/evm2wasm/blob/master/wasm/mod_320.wast
+	if iszero320(y1, y2, y3, y4, y5) {
+		r1 := 0
+		r2 := 0
+		r3 := 0
+		r4 := 0
+		r5 := 0
+	}
```
Comment (axic) [inline libyul/backends/wasm/EVMToEwasmTranslator.cpp:345]: And here.

File: libyul/backends/wasm/EVMToEwasmTranslator.cpp [hunk 4]
```diff
@@ -124,36 +151,338 @@ function mul_128x128_256(x1, x2, y1, y2) -> r1, r2, r3, r4 {
...
+	if iszero512(y1, y2, y3, y4, y5, y6, y7, y8) {
+		r1 := 0
+		r2 := 0
+		r3 := 0
+		r4 := 0
+		r5 := 0
+		r6 := 0
+		r7 := 0
+		r8 := 0
+	}
```
Comment (axic) [inline libyul/backends/wasm/EVMToEwasmTranslator.cpp:380]: Here too.

File: libyul/backends/wasm/EVMToEwasmTranslator.cpp [hunk 5]
```diff
@@ -323,20 +734,55 @@ function shr(x1, x2, x3, x4, y1, y2, y3, y4) -> z1, z2, z3, z4 {
 	}
 }
 function sar(x1, x2, x3, x4, y1, y2, y3, y4) -> z1, z2, z3, z4 {
-	// TODO implement
-	unreachable()
+	let lz := i64.clz(y1)
+	if lz {
+		z1, z2, z3, z4 := shr(x1, x2, x3, x4, y1, y2, y3, y4)
+	}
```
Comment (axic) [inline libyul/backends/wasm/EVMToEwasmTranslator.cpp:719]: And here.

File: libyul/backends/wasm/EVMToEwasmTranslator.cpp [hunk 6]
```diff
@@ -124,36 +151,338 @@ function mul_128x128_256(x1, x2, y1, y2) -> r1, r2, r3, r4 {
 
 	r1 := i64.add(i64.add(ah, carry1), carry2)
 }
+// Multiplies two 256 bit values resulting in a 512 bit
+// value split into eight 64 bit values.
+function mul_256x256_512(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4, r5, r6, r7, r8 {
+	let a1, a2, a3, a4 := mul_128x128_256(x1, x2, y1, y2)
+	let         b1, b2, b3, b4 := mul_128x128_256(x1, x2, y3, y4)
+	let         c1, c2, c3, c4 := mul_128x128_256(x3, x4, y1, y2)
+	let                 d1, d2, d3, d4 := mul_128x128_256(x3, x4, y3, y4)
```
Comment (axic) [inline libyul/backends/wasm/EVMToEwasmTranslator.cpp]: Why are these indented?

**Comment by rodiazet** [comment]
@axic Sorry that It took so long but it seems that I forgot to push the changes to repo.

**Review by chriseth — APPROVED** [review]

**Pull request merged**